File: circuitpython-code/drive.py
"""
The MIT License (MIT)

Copyright (c) 2020 Dave Astels

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

--------------------------------------------------------------------------------
Drive motor abstraction
"""

import board
from digitalio import DigitalInOut, Direction
from pulseio import PWMOut
from adafruit_motor.motor import DCMotor

import adafruit_logging as logging
logger = logging.getLogger('mouse')

STOPPED = 0
FORWARD = 1
REVERSE = 2
PIVOT_LEFT = 3
PIVOT_RIGHT = 4
VEER_LEFT = 5
VEER_RIGHT = 6

class AbstractDriveSystem(object):

    def __init__(self):
        self._state = STOPPED


    @property
    def state(self):
        return self._state


    def stop(self):
        self._state = STOPPED


    def go_forward(self):
        self._state = FORWARD


    def go_backward(self):
        self._state = REVERSE


    def pivot_left(self):
        self._state = PIVOT_LEFT


    def pivot_right(self):
        self._state = PIVOT_RIGHT


    def veer_left(self):
        self._state = VEER_LEFT


    def veer_right(self):
        self._state = VEER_RIGHT



class FakeDriveSystem(AbstractDriveSystem):

    def __init__(self):
        super(FakeDriveSystem, self).__init__()

    def enable(self):
        logger.info('Drive: Enabling motors')


    def disable(self):
        logger.info('Drive: Disabling motors')


    def stop(self):
        super(FakeDriveSystem, self).stop()
        logger.info('Drive: Stop')


    def go_forward(self):
        super(FakeDriveSystem, self).go_forward()
        logger.info('Drive: Forward')


    def go_backward(self):
        super(FakeDriveSystem, self).go_backward()
        logger.info('Drive: Reverse')


    def pivot_left(self):
        super(FakeDriveSystem, self).pivot_left()
        logger.info('Drive: Pivot left')


    def pivot_right(self):
        super(FakeDriveSystem, self).pivot_right()
        logger.info('Drive: Pivot right')


    def veer_left(self):
        super(FakeDriveSystem, self).veer_left()
        logger.info('Drive: Veer left')


    def veer_right(self):
        super(FakeDriveSystem, self).veer_right()
        logger.info('Drive: Veer right')




class DriveSystem(AbstractDriveSystem):

    def __init__(self):
        super(DriveSystem, self).__init__()
        self._motor_enable = DigitalInOut(board.A2)
        self._motor_enable.direction = Direction.OUTPUT
        self._motor_enable.value = False

        self._right_motor = DCMotor(PWMOut(board.D12), PWMOut(board.D13))
        self._left_motor = DCMotor(PWMOut(board.D4), PWMOut(board.D11))

        self._base_speed = 0.5


    @property
    def base_speed(self):
        return self._base_speed

    @base_speed.setter
    def base_speed(self, value):
        self._base_speed = value


    def enable(self):
        self._motor_enable.value = True


    def disable(self):
        self._motor_enable.value = False


    def stop(self):
        super(DriveSystem, self).stop()
        self._left_motor.throttle = 0
        self._right_motor.throttle = 0


    def go_forward(self):
        super(DriveSystem, self).go_forward()
        self._left_motor.throttle = self._base_speed
        self._right_motor.throttle = self._base_speed


    def go_backward(self):
        super(DriveSystem, self).go_backward()
        self._left_motor.throttle = -self._base_speed
        self._right_motor.throttle = -self._base_speed


    def pivot_left(self):
        super(DriveSystem, self).pivot_left()
        self._left_motor.throttle = -0.5 * self._base_speed
        self._right_motor.throttle = 0.5 * self._base_speed


    def pivot_right(self):
        super(DriveSystem, self).pivot_right()
        self._left_motor.throttle = 0.5 * self._base_speed
        self._right_motor.throttle = -0.5 * self._base_speed


    def veer_left(self):
        super(DriveSystem, self).veer_left()
        self._left_motor.throttle = 0.5 * self._base_speed
        self._right_motor.throttle = self._base_speed


    def veer_right(self):
        super(DriveSystem, self).veer_right()
        self._left_motor.throttle = self._base_speed
        self._right_motor.throttle = 0.5 * self._base_speed



def make_drive(debug):
    if debug:
        return FakeDriveSystem()
    else:
        return DriveSystem()

File: c++-code/src/drive.cpp
// Motor interface
//
// Copyright (c) 2020 Dave Astels

#include <Arduino.h>
#include "logging.h"
#include "drive.h"
#include "system.h"

extern Logger *logger;


// -----------------------------------------------------------------------------
// Abstract base

Drive::Drive(System *sys)
  : _state(DriveState::STOPPED)
  , _system(sys)
{
}


Drive *Drive::make_drive(bool debug_mode, System *sys)
{
  if (debug_mode) {
    return new FakeDrive(sys);
  } else {
    return new RealDrive(sys);
  }
}


void Drive::stop()
{
  if (_state == DriveState::STOPPED) {
    return;
  }
  _state = DriveState::STOPPED;
  logger->debug("Drive stopped");
}


void Drive::forward()
{
  if (_state == DriveState::FORWARD) {
    return;
  }
  _state = DriveState::FORWARD;
  logger->debug("Drive forward");
}


void Drive::reverse()
{
  if (_state == DriveState::REVERSE) {
    return;
  }
  _state = DriveState::REVERSE;
  logger->debug("Drive reverse");
}


void Drive::pivot_left()
{
  if (_state == DriveState::PIVOT_LEFT) {
    return;
  }
  _state = DriveState::PIVOT_LEFT;
  logger->debug("Drive pivot left");
}


void Drive::pivot_right()
{
  if (_state == DriveState::PIVOT_RIGHT) {
    return;
  }
  _state = DriveState::PIVOT_RIGHT;
  logger->debug("Drive pivot right");
}


void Drive::veer_left()
{
  if (_state == DriveState::VEER_LEFT) {
    return;
  }
  _state = DriveState::VEER_LEFT;
  logger->debug("Drive veer left");
}


void Drive::veer_right()
{
  if (_state == DriveState::VEER_RIGHT) {
    return;
  }
  _state = DriveState::VEER_RIGHT;
  logger->debug("Drive veer right");
}


// -----------------------------------------------------------------------------
// Fake implementation for desk debugging

FakeDrive::FakeDrive(System *sys)
  :Drive(sys)
{
}


void FakeDrive::stop()
{
  Drive::stop();
}


void FakeDrive::forward()
{
  Drive::forward();
}


void FakeDrive::reverse()
{
  Drive::reverse();
}


void FakeDrive::pivot_left()
{
  Drive::pivot_left();
}


void FakeDrive::pivot_right()
{
  Drive::pivot_right();
}


void FakeDrive::veer_left()
{
  Drive::veer_left();
}


void FakeDrive::veer_right()
{
  Drive::veer_right();
}


// -----------------------------------------------------------------------------
// Real implementation using the motors

RealDrive::RealDrive(System *sys)
  : Drive(sys)
  , _motor_enable(new DigitalOutput(16))
  , _left(new Motor(0, 8, 9, 7, 28)) // motor 1, encoder on Grove 7
  , _right(new Motor(1, 10, 11, 0, 1)) // motor 2, encoder on Grove 1
  , _base_speed(128)
  , _half_speed(64)
  , _indicator(sys->indicator())
{
}


void RealDrive::base_speed(int speed)
{
  Drive::base_speed(speed);
  _base_speed = speed;
  _half_speed = speed / 2;
}


void RealDrive::enable()
{
  _motor_enable->set();
}


void RealDrive::disable()
{
  _motor_enable->clear();
}


void RealDrive::stop()
{
  Drive::stop();
  _left->throttle(0);
  _right->throttle(0);
  _indicator->left_drive(0);
  _indicator->right_drive(0);
}


void RealDrive::forward()
{
  Drive::forward();
  _left->throttle(_base_speed);
  _right->throttle(_base_speed);
  _indicator->left_drive(_base_speed);
  _indicator->right_drive(_base_speed);
}


void RealDrive::reverse()
{
  Drive::reverse();
  _left->throttle(-_base_speed);
  _right->throttle(-_base_speed);
  _indicator->left_drive(-_base_speed);
  _indicator->right_drive(-_base_speed);
}


void RealDrive::pivot_left()
{
  Drive::pivot_left();
  _left->throttle(-_half_speed);
  _right->throttle(_half_speed);
  _indicator->left_drive(-_half_speed);
  _indicator->right_drive(_half_speed);
}


void RealDrive::pivot_right()
{
  Drive::pivot_right();
  _left->throttle(_half_speed);
  _right->throttle(-_half_speed);
  _indicator->left_drive(_half_speed);
  _indicator->right_drive(-_half_speed);
}


void RealDrive::veer_left()
{
  Drive::veer_left();
  _left->throttle(_half_speed);
  _right->throttle(_base_speed);
  _indicator->left_drive(_half_speed);
  _indicator->right_drive(_base_speed);
}


void RealDrive::veer_right()
{
  Drive::veer_right();
  _left->throttle(_base_speed);
  _right->throttle(_half_speed);
  _indicator->left_drive(_base_speed);
  _indicator->right_drive(_half_speed);
}
